src/python/turicreate/_deps/__init__.py

```diff
@@ -9,59 +9,346 @@
+from types import ModuleType
 
+if six.PY2:
+    import imp
+else:
+    import importlib
+
+def run_from_ipython():
+    # Taken from https://stackoverflow.com/questions/5376837
+    try:
+        __IPYTHON__
+        return True
+    except NameError:
+        return False
+
+## builtin __import__(path, globals, locals, fromlist)
+# referenced from https://github.com/mnmelo/lazy_import/blob/master/lazy_import/__init__.py
+if six.PY2:
+    import imp
+    # reentrant lock
+    # https://docs.python.org/2/library/imp.html#imp.acquire_lock
+    class _ImportLockContext:
+        def __enter__(self):
+            imp.acquire_lock()
+        def __exit__(self, exc_type, exc_value, exc_traceback):
+            imp.release_lock()
+else:
+    # importlib uses _imp internally; check source you will know ;-)
+    from importlib._bootstrap import _ImportLockContext
+
+# should always be guared by _ImportLockContext
+def default_init_func(name, *args, **kwargs):
+    """
+    called within _ImportLockContext(); no import lock should be hold inside
+    """
+    _ret = sys.modules.get(name, None)
+    if _ret is not None:
+        assert name not in sys.modules.keys(), ("sys.modules[%s] cannot be None"
+                                                " during module loading" % name)
+    if six.PY2:
+        # imp importing submodule is too tricky
+        # moulde and pkg naming should follow pep8
+        if not all([x.isalnum() or x in '._' for x in name]):
+            raise ValueError('invalid module name')
+        exec("import %s as _py2_ret" % name)
+        return _py2_ret
+    else:
+        return importlib.import_module(name)
+
+class LazyModuleLoader(ModuleType, object):
+    """ defer to load a module. If it's loaded, no exception but warning will be given. """
+    # read-only member list -> used by __setattr__; see below
+    _my_attrs_ = ['is_loaded', 'reload', 'get_module',
+                  '_name', '_init_func', '_module', '_loaded']
+    # ensure singleton; must be guarded by _ImportLockContext lock
+    registry_ = set()
+
+    def __init__(self, name, init_func=default_init_func, is_lambda=False):
+        """
+        Once write then read only. `reload` can reset the state.
+        Only singleton instance is allowed.
+        Context _ImportLockContext is reentrant.
+
+        This Class only supports syntax `import a.b [as c]`;
+        you need to translate `from a import b as c` to the form described above.
+
+        LazyModuleLoader hijacks sys.modules[name] to defer the
+        use of the module if the module is not already loaded yet.
+
+        The load of module is deferred until LazyModuleLoader is called
+        by __getattr__, __setattr__, or __delattr__ on `attr` other than
+        `dir(LazyModuleLoader())` The side effect is that sys.modules[name]
+        will be set to real module object.
+
+        >>> np = LazyModuleLoader('numpy')
+        >>> import numpy # won't actually load
```

Comment:
I don't understand why this is the case...  To do this, it would actually have to do injection into sys.modules, which I don't see happening below.

Reply:
Good catch! That's the previous behavior. I will hijack sys.modules['name'].